ftp/src/main/scala/akka.stream.contrib.ftp/FtpSource.scala

```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright (C) 2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+package akka.stream.contrib.ftp
+
+import java.net.InetAddress
+import akka.stream.{ Attributes, Outlet, SourceShape }
+import akka.stream.Attributes.{ InputBuffer, name }
+import akka.stream.contrib.ftp.FtpConnectionSettings.{ BasicFtpConnectionSettings, DefaultFtpConnectionSettings, DefaultFtpPort }
+import akka.stream.contrib.ftp.FtpCredentials.{ AnonFtpCredentials, NonAnonFtpCredentials }
+import akka.stream.scaladsl.Source
+import akka.stream.stage.{ GraphStageLogic, GraphStageWithMaterializedValue, OutHandler }
+import org.apache.commons.net.ftp.{ FTPClient, FTPListParseEngine }
+import scala.concurrent.{ Future, Promise }
+import scala.util.control.NonFatal
+
+object FtpSource {
+
+  final val SourceName = "ftpSource"
+
+  def apply(): Source[FtpFile, Future[Long]] = apply(DefaultFtpConnectionSettings)
+
+  def apply(hostname: String): Source[FtpFile, Future[Long]] = apply(hostname, DefaultFtpPort)
+
+  def apply(hostname: String, port: Int): Source[FtpFile, Future[Long]] =
+    apply(BasicFtpConnectionSettings(InetAddress.getByName(hostname), port, AnonFtpCredentials))
+
+  def apply(hostname: String, username: String, password: String): Source[FtpFile, Future[Long]] =
+    apply(hostname, DefaultFtpPort, username, password)
+
+  def apply(hostname: String, port: Int, username: String, password: String): Source[FtpFile, Future[Long]] =
+    apply(BasicFtpConnectionSettings(InetAddress.getByName(hostname), port, NonAnonFtpCredentials(username, password)))
+
+  def apply(connectionSettings: FtpConnectionSettings): Source[FtpFile, Future[Long]] =
+    Source.fromGraph(new FtpSource(connectionSettings)).withAttributes(name(SourceName))
+
+}
+
+final class FtpSource private (connectionSettings: FtpConnectionSettings)
+  extends GraphStageWithMaterializedValue[SourceShape[FtpFile], Future[Long]] {
+  import FtpSource._
+
+  val matValue = Promise[Long]()
+
+  val shape = SourceShape(Outlet[FtpFile](s"$SourceName.out"))
+
+  def createLogicAndMaterializedValue(inheritedAttributes: Attributes) = {
+
+    val InputBuffer(initialBuffer, maxBuffer) =
+      inheritedAttributes.getAttribute(classOf[InputBuffer], InputBuffer(16, 16))
```

Comment:
was iffy on the buffer, since we already have buffers in the infrastructure but I guess prefetch makes sense since ftp fetching may be expensive i guess